aip/0234.md

```diff
@@ -0,0 +1,83 @@
+---
+aip:
+  id: 234
+  state: draft
+  created: 2019-06-18
+permalink: /234
+redirect_from:
+  - /0234
+---
+
+# Batch Update
+
+Some APIs need to allow users to modify a set of resources in a single
+transaction. A batch update method provides this functionality.
+
+## Guidance
+
+Batch update methods are specified using the following pattern:
+
+```proto
+rpc BatchUpdateBooks(BatchUpdateBooksRequest) returns (BatchUpdateBooksResponse) {
+  option (google.api.http) = {
+    post: "/v1/shelves/-/books:batchUpdate"
+    body: "*"
+  };
+}
+```
+
+- The RPC's name must begin with `BatchUpdate`. The remainder of the RPC name
+  should be the plural form of the resource being updated.
+- The request and response messages **must** match the RPC name, with `-Request`
+  and `-Response` suffixes.
+  - However, in the event that the request may take a significant amount of
+    time, the response message **must** be a `google.longrunning.Operation`
+    which ultimately resolves to the `-Response` type.
+- The HTTP verb must be `POST`.
+- The HTTP URI must end with `:batchUpdate`.
+- The URL path **should** represent the collection for the resource, matching
+  the collection used for simple CRUD operations. If the operation spans
+  parents, a dash (`-`) **may** be used as a wildcard.
+- The body clause in the `google.api.http` annotation should be `"*"`.
+- The operation should be atomic: it should fail for all resources or succeed
+  for all resources (no partial success).
+
+### Request message
+
+The request for a batch update method should be specified with the following
+pattern:
+
+```proto
+message BatchUpdateBookRequest {
+  // The request message specifying the resources to update.
+  // A maximum of 1000 books can be modified in a batch.
+  repeated UpdateBookRequest requests = 1;
```

Comment:
An alternative would be:

```proto
repeated Book books = 1;
google.protobuf.FieldMask update_mask = 2;
```

This has one fewer layer of indirection, but carries with it the restriction that you have to have the same field mask across the board. (Personally, I think that restriction is probably a feature, but perhaps not?)

Reply:
I think I have a slight preference for removing the indirection too. If callers are doing wildly disparate updates or creates across parents, to get around quota/billing issues, then we're just redefining what HTTP multipart does, and that means atomicity isn't important and these wrapper messages are needed.

If, however, we're trying to solve for transactional requirements, then atomicity is important, and it's more likely that requests look similar to each other (like update en-mass the same set of fields), and the level of indirection is **not** needed.

So it seems like we should thing about what requirement/need we're trying to solve.

Reply:
(See other discussions on this topic.)